src/ws/cockpitbranding.c

```diff
@@ -0,0 +1,278 @@
+    g_signal_handler_disconnect (data->transport, data->transport_closed);
+  data->transport_closed = 0;
+
+  g_object_unref (data->response);
+  g_object_unref (data->service);
+
+  if (data->local_os_release)
+    g_hash_table_unref (data->local_os_release);
+
+  g_free (data->channel);
+  g_free (data);
+}
+
+static void
+on_init_ready (CockpitBrandingData *data)
+{
+  gchar **roots = NULL;
+  /* owned by cockpit webservice */
+  GHashTable *os_release = cockpit_web_service_get_transport_os_release (data->service,
+                                                                         data->transport);
+
+  if (os_release)
+    {
+      roots = cockpit_branding_calculate_static_roots (g_hash_table_lookup (os_release, "ID"),
+                                                       g_hash_table_lookup (os_release, "VARIANT_ID"),
+                                                       FALSE);
+    }
+  else
+    {
+      roots = cockpit_branding_calculate_static_roots (NULL, NULL, FALSE);
+    }
+
+  serve_branding_css_file (data->response, data->path,
+                           (const gchar **)roots, os_release);
+  g_strfreev (roots);
+  cockpit_branding_data_free (data);
+}
+
+static gboolean
+on_transport_control (CockpitTransport *transport,
+                      const gchar *command,
+                      const gchar *channel,
+                      JsonObject *options,
+                      GBytes *message,
+                      gpointer user_data)
+{
+  CockpitBrandingData *data = user_data;
+  JsonObject *object = NULL;
+  GBytes *bytes = NULL;
+
+  if (!channel || !g_str_equal (channel, data->channel))
+    return FALSE; /* not handled */
+
+  /* Close the channel */
+  object = cockpit_transport_build_json ("command", "close",
+                                         "channel", channel,
+                                         NULL);
+  bytes = cockpit_json_write_bytes (object);
+  cockpit_transport_send (transport, NULL, bytes);
+  g_bytes_unref (bytes);
+  json_object_unref (object);
+
+  on_init_ready (data);
+  return TRUE;
+}
+
+static void
+on_transport_closed (CockpitTransport *transport,
+                     const gchar *problem,
+                     gpointer user_data)
+{
+  CockpitBrandingData *data = user_data;
+  cockpit_web_response_error (data->response, 502, NULL, NULL);
+  cockpit_branding_data_free (data);
+}
+
+static void
+serve_branding_css_with_init_data (CockpitWebService *service,
+                                   CockpitWebResponse *response,
+                                   const gchar *path,
+                                   GHashTable *local_os_release)
+{
+  CockpitTransport *transport = NULL;
+  JsonObject *open = NULL;
+  gchar *channel = NULL;
+  CockpitBrandingData *cbd = NULL;
+  GBytes *bytes = NULL;
+
+  /* Open a channel to ensure we got an init */
+  channel = cockpit_web_service_unique_channel (service);
+  open = cockpit_transport_build_json ("command", "open",
```

Comment:
I was trying to avoid having to maintain a list of pending callbacks for a CockpitSession in the webservice but if you think that's better I'll do that instead of waiting for a channel.